src/hotspot/share/memory/archiveBuilder.cpp

```diff
@@ -593,6 +595,226 @@ void ArchiveBuilder::make_klasses_shareable() {
+    top = runtime_base + size;
+    log_info(cds, map)("[%-18s " PTR_FORMAT " - " PTR_FORMAT " " SIZE_FORMAT_W(9) " bytes]",
+                       name, p2i(base), p2i(top), size);
+  }
+
+  // open and closed archive regions
+  void write_heap_region(const char* which, GrowableArray<MemRegion> *regions) {
+    for (int i = 0; i < regions->length(); i++) {
+      address start = address(regions->at(i).start());
+      address end = address(regions->at(i).end());
+      write_region(which, start, end, start);
+      write_data(start, end, start);
+    }
+  }
+
+  // Dump all the data [base...top). Pretend that the base address
+  // will be mapped to runtime_base at run-time.
+  static void write_data(address base, address top, address runtime_base) {
+    assert(top >= base, "must be");
+
+    if (log_is_enabled(Trace, cds, map)) {
+      intx* p = (intx*)base;
+      size_t words = (top - base) / sizeof(uintx);
+
+      for (size_t i = 0; i < words; i += 4, p += 4) {
+        size_t offset = address(p) - base;
+        address runtime_p = runtime_base + offset;
+        switch (words - i) {
+        case 1:
+          log_trace(cds, map)(PTR_FORMAT ": " PTR_FORMAT,
+                              p2i(runtime_p), p[0]);
+          break;
+        case 2:
+          log_trace(cds, map)(PTR_FORMAT ": " PTR_FORMAT " " PTR_FORMAT,
+                              p2i(runtime_p), p[0], p[1]);
+          break;
+        case 3:
+          log_trace(cds, map)(PTR_FORMAT ": " PTR_FORMAT " " PTR_FORMAT " " PTR_FORMAT,
+                              p2i(runtime_p), p[0], p[1], p[2]);
+          break;
+        default:
+          log_trace(cds, map)(PTR_FORMAT ": " PTR_FORMAT " " PTR_FORMAT " " PTR_FORMAT " " PTR_FORMAT,
+                              p2i(runtime_p), p[0], p[1], p[2], p[3]);
+        }
+      }
+    }
+  }
+
+public:
+  CDSMapLogger() {}
+  void write(ArchiveBuilder* builder, FileMapInfo* mapinfo,
+             GrowableArray<MemRegion> *closed_heap_regions,
+             GrowableArray<MemRegion> *open_heap_regions,
+             char* bitmap, size_t bitmap_size_in_bytes) {
+    log_info(cds, map)("%s CDS archive map for %s", DumpSharedSpaces ? "Static" : "Dynamic", mapinfo->full_path());
+    if (log_is_enabled(Trace, cds, map)) {
+      log_info(cds, map)("Log level = trace");
+    } else if (log_is_enabled(Debug, cds, map)) {
+      log_info(cds, map)("Log level = debug");
+      log_info(cds, map)("Run with -Xlog:cds+map=trace for more detailed information");
+    } else {
+      log_info(cds, map)("Log level = info");
+      log_info(cds, map)("Run with -Xlog:cds+map=debug for more detailed information");
+      log_info(cds, map)("Run with -Xlog:cds+map=trace for even more detailed information");
+    }
+
+    address header = address(mapinfo->header());
+    address header_end = header + mapinfo->header()->header_size();
+    write_region("header", header, header_end, 0);
+    write_data(header, header_end, 0);
```

Comment:
Consider writing out the header members human-readable, that would be more useful than a hexdump. I'd also print them already at Info level, since often the header info is the only thing interesting.